msu/hooks/skills/skill_container.nut

```diff
@@ -27,6 +27,67 @@
 		this.m.ScheduledChangesSkills.clear();
 	}
 
+	local add = o.add;
+	o.add = function( _skill, _order = 0 )
+	{
+		if (_skill.isStacking()) return add(_skill, _order);
+
+		local incrementAddedStack = function( _alreadyPresentSkill, _skillToAdd )
+		{
+			if (++_alreadyPresentSkill.m.MSU.AddedStack > 0) _alreadyPresentSkill.m.IsGarbage = false;
+
+			if (!::MSU.isNull(_skillToAdd.getItem()))
+			{
+				if (::MSU.isNull(_alreadyPresentSkill.getItem()) _alreadyPresentSkill.setItem(_skillToAdd.getItem());
+				foreach (i, itemSkill in _skillToAdd.getItem().m.SkillPtrs)
+				{
+					if (itemSkill.getID() == _skillToAdd.getID())
+					{
+						_skillToAdd.getItem().m.SkillPtrs[i] = _alreadyPresentSkill;
+						_skillToAdd.setItem(null);
+						break;
+					}
+				}
+			}
+		}
+
+		foreach (alreadyPresentSkill in this.m.Skills)
+		{
+			if (alreadyPresentSkill.getID() == _skill.getID())
+			{
+				incrementAddedStack(alreadyPresentSkill, _skill);
+				if (alreadyPresentSkill.m.MSU.AddedStack > 1) alreadyPresentSkill.onRefresh();
+				return;
+			}
+		}
+
+		foreach (alreadyPresentSkill in this.m.SkillsToAdd)
+		{
+			if (alreadyPresentSkill.getID() == _skill.getID())
+			{
+				incrementAddedStack(alreadyPresentSkill, _skill);
+				return;
+			}
+		}
+	}
```

Comment:
There's a missing right bracket on line 41, but honestly after some more thought I really dislike this function to the point of thinking we should redo it entirely anyway to get rid of that closure.
There's also another mistake I think in that the original `add` isn't being run if this is the first time a stacking skill is being added (there should be a `return add` at the end)
Suggestion for rewrite, untested:
```squirrel
local add = o.add;
o.add = function( _skill, _order = 0 )
{
	if (_skill.isStacking()) return add(_skill, _order); 
	local i = 0;
	local allSkills = clone this.m.Skills
	allSkills.extend(this.m.SkillsToAdd)
	foreach (alreadyPresentSkill in allSkills)
	{
		++i;
		if (alreadyPresentSkill.getID() == _skill.getID())
		{
			if (++alreadyPresentSkill.m.MSU.AddedStack > 0) alreadyPresentSkill.m.IsGarbage = false;
			if (!::MSU.isNull(_skill.getItem()))
			{
				if (::MSU.isNull(alreadyPresentSkill.getItem())) alreadyPresentSkill.setItem(_skill.getItem());
				foreach (i, itemSkill in _skill.getItem().m.SkillPtrs)
				{
					if (itemSkill.getID() == _skill.getID())
					{
						_skill.getItem().m.SkillPtrs[i] = alreadyPresentSkill;
						_skill.setItem(null);
						break;
					}
				}
			}
			if (alreadyPresentSkill.m.MSU.AddedStack > 1 && this.m.Skills.len() <= i) alreadyPresentSkill.onRefresh();
			return;
		}
	}
	return add(_skill, _order);
}
```
I also just generally dislike the fact that there was a syntax error that would've come up during compilation in this PR and I'm gonna ask for some testing to be done to make sure all this works.

Reply:
Yeah sorry the code got some mistakes after I refactored it based on Taro's review. Your rewrite looks like a good approach. I'll fix the thing.